### PR TITLE
Fortune v2

### PR DESCRIPTION
**Pull request by eugenvoronov**
## Description
Refactored version of the Job Launcher, can be used as a boilerplate for other services.

## Summary of changes

- [x] Removed interfaces
- [x] Refactored DTO
- [x] Token module was merged with Auth module
- [x] Repositories was added
- [x] Other small fixes and improvements

## Related issues
[Refactoring Job Launcher service#486](https://github.com/humanprotocol/human-protocol/issues/486)

### CODE REVIEW

**Comment by mrhouzlane** [comment]
Can you add an env.example ?

File: packages/examples/fortune-v2/job-launcher/server/src/migrations/1677867973538-addUserTable.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+              isUnique: true,
+            },
+            {
+              name: "socket_id",
+              type: "varchar",
+              isUnique: true,
+              isNullable: true,
+            },
+            {
+              name: "type",
```
[the remark this answers is not in the document]
Comment (eugenvoronov) [reply packages/examples/fortune-v2/job-launcher/server/src/migrations/1677867973538-addUserTable.ts:57]: Yes, it for future administration module

File: packages/examples/fortune-v2/job-launcher/server/src/main.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  const sessionHash = bcrypt.hashSync(sessionSecrete, sessionSalt);
+
+  console.log(sessionHash)
+
+  app.use(
+    session({
+      secret: sessionHash,
+      resave: false,
+      saveUninitialized: false,
+    }),
```
[the remark this answers is not in the document]
Comment (eugenvoronov) [reply packages/examples/fortune-v2/job-launcher/server/src/main.ts]: fixed with cookie parser

File: packages/examples/fortune-v2/job-launcher/server/src/common/constants/errors.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+export enum Auth {
+  Unauthorized = "Unauthorized",
+}
+
+export enum Job {
```
Comment (mrhouzlane) [inline packages/examples/fortune-v2/job-launcher/server/src/common/constants/errors.ts:5]: Would name it : JobError ?

File: packages/examples/fortune-v2/job-launcher/server/src/common/constants/errors.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  NotFound = "Job not found",
+  NotCreated = "Job has not been created",
+  NotEnoughFunds = "Not enough funds"
+}
+
+export enum Escrow {
```
Comment (mrhouzlane) [inline packages/examples/fortune-v2/job-launcher/server/src/common/constants/errors.ts:11]: EscrowError same here

File: packages/examples/fortune-v2/job-launcher/server/src/common/constants/errors.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+export enum Auth {
+  Unauthorized = "Unauthorized",
+}
+
+export enum Job {
+  NotFound = "Job not found",
+  NotCreated = "Job has not been created",
+  NotEnoughFunds = "Not enough funds"
+}
+
```
Comment (mrhouzlane) [inline packages/examples/fortune-v2/job-launcher/server/src/common/constants/errors.ts:10]: /**
 * Represents escrow-related error messages.
 */

File: packages/examples/fortune-v2/job-launcher/server/src/common/constants/errors.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+export enum Escrow {
+  NotFound = "Escrow not found",
+  NotCreated = "Escrow has not been created"
+}
+
```
Comment (mrhouzlane) [inline packages/examples/fortune-v2/job-launcher/server/src/common/constants/errors.ts:15]: /**
 * Represents user-related error messages.
 */

File: packages/examples/fortune-v2/job-launcher/server/src/common/decorators/user.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import { UserStatus, UserType } from "../enums/user";
+import { IBase } from "./base";
+
+export interface IUser extends IBase {
+  password: string;
+  email: string;
+  status: UserStatus;
+  type: UserType;
+}
+
```
Comment (mrhouzlane) [inline packages/examples/fortune-v2/job-launcher/server/src/common/decorators/user.ts:11]: /**
 * Custom decorator to retrieve the currently logged-in user.
 * Returns the user object from the request context.
 */

File: packages/examples/fortune-v2/job-launcher/server/src/common/decorators/user.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+export const User = createParamDecorator((_data: unknown, ctx: ExecutionContext) => {
+  const request = ctx.switchToHttp().getRequest();
```
Comment (mrhouzlane) [inline packages/examples/fortune-v2/job-launcher/server/src/common/decorators/user.ts:13]: maybe also handling non Http contexts, like GraphQL or Websockets

File: packages/examples/fortune-v2/job-launcher/server/src/common/dynamic/index.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+export interface IAsyncModuleConfig<T> extends Pick<ModuleMetadata, "imports" | "exports"> {
+  useExisting?: {
+    value: IModuleConfigFactory<T>;
+    provide?: InjectionToken;
+  };
+  useClass?: Type<IModuleConfigFactory<T>>;
+  useFactory?: (...args: any[]) => Promise<T> | T;
+  inject?: any[];
+}
+
```
Comment (mrhouzlane) [inline packages/examples/fortune-v2/job-launcher/server/src/common/dynamic/index.ts:21]: /**
 * Creates a module configuration provider based on the provided options.
 * @param provide - The injection token for the module configuration.
 * @param options - The options for creating the module configuration provider.
 * @returns An array of providers for the module configuration.
 */

File: packages/examples/fortune-v2/job-launcher/server/src/common/guards/jwt.http.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+import { Reflector } from "@nestjs/core";
+import { AuthGuard } from "@nestjs/passport";
+
+@Injectable()
+export class JwtHttpGuard extends AuthGuard("jwt-http") implements CanActivate {
+  constructor(private readonly reflector: Reflector) {
+    super();
+  }
+
+  public async canActivate(context: ExecutionContext): Promise<boolean> {
```
Comment (mrhouzlane) [inline packages/examples/fortune-v2/job-launcher/server/src/common/guards/jwt.http.ts:11]: better use try/catch with async (suggestion)

**Review by mrhouzlane — COMMENTED** [review]
- Posted some optional suggestions

**Review by portuu3 — DISMISSED** [review]

File: packages/examples/fortune-v2/job-launcher/server/src/common/decorators/user.ts (hunk 8 above)
Comment (eugenvoronov) [reply packages/examples/fortune-v2/job-launcher/server/src/common/decorators/user.ts:13]: Lets leave it as improvements, I will add it to backlog

**Review by portuu3 — APPROVED** [review]

**Pull request merged**